Description: ubuntu xend probe
 .
 libvirt (1.2.8-0ubuntu2) utopic; urgency=low
 .
  * d/p/ubuntu-xend-probe.patch:
    Update patch correctly and re-enable it. It seems like it only was
    half updated and then disabled without reasons.
Author: Stefan Bader <stefan.bader@canonical.com>
Last-Update: 2015-11-24

Index: libvirt-1.3.3/src/libxl/libxl_driver.c
===================================================================
--- libvirt-1.3.3.orig/src/libxl/libxl_driver.c
+++ libvirt-1.3.3/src/libxl/libxl_driver.c
@@ -442,6 +442,9 @@ static bool
 libxlDriverShouldLoad(bool privileged)
 {
     bool ret = false;
+    int status;
+    char *output = NULL;
+    virCommandPtr cmd;
 
     /* Don't load if non-root */
     if (!privileged) {
@@ -450,8 +453,6 @@ libxlDriverShouldLoad(bool privileged)
     }
 
     if (virFileExists(HYPERVISOR_CAPABILITIES)) {
-        int status;
-        char *output = NULL;
         /*
          * Don't load if not running on a Xen control domain (dom0). It is not
          * sufficient to check for the file to exist as any guest can mount
@@ -474,19 +475,23 @@ libxlDriverShouldLoad(bool privileged)
     }
 
     /* Don't load if legacy xen toolstack (xend) is in use */
-    if (virFileExists("/usr/sbin/xend")) {
-        virCommandPtr cmd;
+    cmd = virCommandNewArgList("/usr/lib/xen-common/bin/xen-toolstack", NULL);
+    virCommandSetOutputBuffer(cmd, &output);
+    if (virCommandRun(cmd, &status) == 0 && status == 0) {
+        int i, j;
+
+        for (i = 0, j = 0; output[i] != '\0'; i++)
+            if (output[i] == '/')
+                j = i + 1;
 
-        cmd = virCommandNewArgList("/usr/sbin/xend", "status", NULL);
-        if (virCommandRun(cmd, NULL) == 0) {
+        if (output[j] == 'x' && output[j+1] == 'l') {
+            ret = true;
+        } else {
             VIR_INFO("Legacy xen tool stack seems to be in use, disabling "
                      "libxenlight driver.");
-        } else {
-            ret = true;
         }
+        VIR_FREE(output);
         virCommandFree(cmd);
-    } else {
-        ret = true;
     }
 
     return ret;
Index: libvirt-1.3.3/src/xen/xen_driver.c
===================================================================
--- libvirt-1.3.3.orig/src/xen/xen_driver.c
+++ libvirt-1.3.3/src/xen/xen_driver.c
@@ -300,22 +300,36 @@ xenUnifiedProbe(void)
 }
 
 #ifdef WITH_LIBXL
+#define TOOLSTACK_PATH "/usr/lib/xen-common/bin/xen-toolstack"
 static bool
 xenUnifiedXendProbe(void)
 {
     bool ret = false;
+    char *output;
 
-    if (virFileExists("/usr/sbin/xend")) {
+    if (virFileExists(TOOLSTACK_PATH)) {
         virCommandPtr cmd;
+        int status;
 
-        cmd = virCommandNewArgList("/usr/sbin/xend", "status", NULL);
-        if (virCommandRun(cmd, NULL) == 0)
-            ret = true;
+        cmd = virCommandNewArgList(TOOLSTACK_PATH,  NULL);
+        virCommandSetOutputBuffer(cmd, &output);
+        if (virCommandRun(cmd, &status) == 0 && status == 0) {
+            int i, j;
+
+            for (i = 0, j = 0; output[i] != '\0'; i++)
+                if (output[i] == '/')
+                    j = i + 1;
+
+            if (output[j] == 'x' && output[j+1] == 'm')
+                ret = true;
+        }
+        VIR_FREE(output);
         virCommandFree(cmd);
     }
 
     return ret;
 }
+#undef TOOLSTACK_PATH
 #endif
 
 
